### PR TITLE
Java/C++/C#: Refactor dataflow to simplify return flow.

### PR DESCRIPTION
**Pull request by aschackmull**
This merges the two kinds of return flow (value flow from `ReturnNode` to `OutNode` and parameter updates through `PostUpdateNode`s) to reduce the number of cases in the data flow implementation.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll [hunk 1]
```diff
@@ -736,35 +736,102 @@ private module ImplCommon {
...
+      or
+      exists(ParameterNode p, int pos |
+        parameterValueFlowsToUpdate(p, this) and
+        p.isParameterOf(_, pos) and
+        result = TParamUpdate(pos)
+      )
+    }
+  }
+
+  private newtype TReturnKindExt =
```
Comment (hvitved) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll]: I think we should cache this, like we cache e.g. `TReturnPosition`.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll [hunk 2]
```diff
@@ -736,35 +736,102 @@ private module ImplCommon {
...
   pragma[noinline]
-  ReturnPosition getReturnPosition(ReturnNode ret) {
-    exists(DataFlowCallable c, ReturnKind k | returnPosition(ret, c, k) |
+  ReturnPosition getReturnPosition(ReturnNodeExt ret) {
+    exists(DataFlowCallable c, ReturnKindExt k | returnPosition(ret, c, k) |
       result = TReturnPosition0(c, k)
     )
   }
 
+  Node getAnOutNodeExt(DataFlowCall call, ReturnKindExt kind) {
```
Comment (hvitved) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll]: Make this an abstract predicate defined by virtual dispatch  on `ReturnKindExt`?

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll [hunk 3]
```diff
@@ -258,8 +258,8 @@ private predicate additionalJumpStep(Node node1, Node node2, Configuration confi
 private predicate useFieldFlow(Configuration config) { config.fieldFlowBranchLimit() >= 1 }
 
 pragma[noinline]
-private ReturnPosition viableReturnPos(DataFlowCall call, ReturnKind kind) {
-  viableImpl(call) = result.getCallable() and
+private ReturnPosition viableReturnPos(DataFlowCall call, ReturnKindExt kind) {
+  viableCallable(call) = result.getCallable() and
```
Comment (hvitved) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll]: This means we no longer need the `viableImpl` predicate (right?). If so, I would like to have [viableImpl](https://github.com/Semmle/ql/blob/master/csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowDispatch.qll#L110) renamed to `viableCallable`, and remove [this](https://github.com/Semmle/ql/blob/master/csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowDispatch.qll#L185) alias. Perhaps a similar simplification is relevant for other languages.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll [hunk 4]
```diff
@@ -736,35 +736,102 @@ private module ImplCommon {
     else result instanceof LocalCallContextAny
   }
 
+  /**
+   * A node from which flow can return to the caller. This is either a regular
+   * `ReturnNode` or a `PostUpdateNode` corresponding to the value of a parameter.
+   */
+  class ReturnNodeExt extends Node {
```
Comment (hvitved) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll]: Should we also have an `OutNodeExt`?

**Comment by hvitved** [comment]
For reference, [here](https://git.semmle.com/gist/tom/2e4feb7c160bd1021193349d08fb58e4) is the dist-compare report for C# (performance is unchanged).

**Comment by yh-semmle** [comment]
There may be a performance issue for Java. One Analyze-Snapshots job (# 4177) timed out, and the retry (# 4193) currently appears to be hanging at the same query. @aschackmull please check this using a JDK 11 snapshot (available from the corresponding Build-Snapshots job # 4837).

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll (hunk 4 above)
Comment (aschackmull) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll]: I don't think that's necessary.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll (hunk 3 above)
Comment (aschackmull) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll]: Yes.  Let's do that in a followup PR.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll (hunk 1 above)
Comment (aschackmull) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll]: I haven't seen any need for this as it's a fairly light-weight predicate - indeed perhaps `TReturnPosition` shouldn't be cached either.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll (hunk 2 above)
Comment (aschackmull) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll]: Hmm, I guess it doesn't really matter much, so sure I could do that.  I guess it should then simply be named `getAnOutNode` as the `Ext` part becomes implicit?

**Comment by aschackmull** [comment]
> There may be a performance issue for Java

Yes, there was some bad magic, which caused a bad join-order in a standard order delta (standard orders pop up to hurt us yet again).  Fixed.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll (hunk 2 above)
Comment (hvitved) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll]: 👍

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll (hunk 3 above)
Comment (hvitved) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll]: OK.

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll (hunk 2 above)
Comment (aschackmull) [reply cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImplCommon.qll]: Done.  Checking one dataflow query, the DIL was essentially unchanged, so I don't expect any perf surprises from this.

**Comment by yh-semmle** [comment]
The Java timeout on JDK 11 has been resolved, but there is a non-negligible increase in overall analysis time for JDK 11 from around 88min to 106min, consistent over two runs. Profile results (from Analyze-Snapshots/4405) suggest that several security queries (incl. ConditionalByPass.ql, ZipSlip.ql, XXE.ql and ImproperValidationOfArrayConstruction.ql) now take noticeably longer, up to about 2x.

@aschackmull could you please take another look to see if this can be improved.

**Comment by aschackmull** [comment]
Testing ZipSlip on jdk11 revealed two issues - one case of bad magic and one bad join-order (which mostly were in a standard delta order, but manifested as a loop invariant join_rhs).  Both are fixed.

**Comment by aschackmull** [comment]
Testing the other 3 queries locally did not uncover additional performance issues.

**Comment by yh-semmle** [comment]
Performance looks good now. There are 3 new results for `java/user-controlled-bypass` (see Java-Differences # 425), is that expected?

**Comment by jbj** [comment]
There has been several surprises in this PR. Shall we delay it until Monday so it doesn't get into 1.23?

**Comment by jbj** [comment]
1.23 has branched off just now.

**Comment by aschackmull** [comment]
> There are 3 new results for java/user-controlled-bypass (see Java-Differences # 425), is that expected?

Those additional results are due to the tiny change in `DataFlowUtil.qll`. Normally you wouldn't expect this to change results, but that query is so noisy on jdk11 that it picks up on very minor changes like this.  Essentially a small precision improvement to an early pruning step can bring more dispatch steps under the threshold for consideration for field flow - this is in most cases not a very relevant effect, but it can be observed through very noisy queries.

**Comment by aschackmull** [comment]
(I just verified that removing the change to `DataFlowUtil.qll` causes the `java/user-controlled-bypass` results to be unchanged on jdk11 by this PR)

**Comment by yh-semmle** [comment]
Thanks for checking, that sounds fine then. The only additional change after reviews from the other languages was 106b8cfb, so this looks good to merge now that 1.23 has been branched off.

**Pull request merged**